### PR TITLE
refactor search reconciling logic

### LINKED ISSUE #3669: registries not reconciled when clusters changed (add, update, del)
A `ResourceRegistry` like below, works, and it should be:

```yaml
apiVersion: search.karmada.io/v1alpha1
kind: ResourceRegistry
metadata:
  name: default-global-search
spec:
  targetCluster:
    labelSelector:
      matchExpressions:
      - key: karmada-cluster.k8s.some.com/managed
        operator: Exists 
  resourceSelectors:
  - apiVersion: v1
    kind: Pod
  - apiVersion: events.k8s.io/v1
    kind: Events
```

It intended to act as a default global registry. However, if a new cluster joined to karmada with a matching label, nothing can be retrieved through search api.

Here, the cluster handler of search controller, test if cluster registry exist, if false, it just returned, without checking labels with any exist registries' label selector:

https://github.com/karmada-io/karmada/blob/4e4bfffc61067ddbde6b287cdb338bea2f5647df/pkg/search/controller.go#L166-L172

**What would you like to be added**:

Clusters add, updated (label add, remove), could trigger reconcile with matching registries.

**Why is this needed**:

By design.

### PR DESCRIPTION
**Pull request by tedli**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

1. clusters change didn't trigger registries reconcile.
2. a cluster may has more than one cluster manager informer.

**Which issue(s) this PR fixes**:
Fixes #3669

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Review by whitewindmills — COMMENTED** [review]
Kindly add e2e case to cover this scenario.

File: pkg/search/controller.go [hunk 1]
```diff
@@ -161,46 +162,219 @@ func (c *Controller) handleErr(err error, key interface{}) {
...
+	cluster := cls.Name
+	var allRegistries map[string]*searchv1alpha1.ResourceRegistry
+	var matched []string
+	allRegistries, matched, err = c.getClusterMatchedRegistries(cls)
+	if err != nil {
+		klog.Errorf("get cluster matched resource registries failed, cluster: %s, error: %s", cluster, err)
+		return
+	}
+	if len(matched) > 0 {
+		klog.V(4).Infof("registries matched with cluster, cluster: %s, registries: %s",
```
Comment (whitewindmills) [inline pkg/search/controller.go:278]: It is recommended to keep the first letter of all log printing in uppercase.

File: pkg/util/fedinformer/genericmanager/multi-cluster-manager.go [hunk 2]
```diff
@@ -145,3 +148,13 @@ func (m *multiClusterInformerManagerImpl) WaitForCacheSyncWithTimeout(cluster st
 	}
 	return manager.WaitForCacheSyncWithTimeout(cacheSyncTimeout)
 }
+
+func (m *multiClusterInformerManagerImpl) String() string {
+	if m.managers == nil {
+		return "nil"
+	}
+	if len(m.managers) <= 0 {
```
Comment (whitewindmills) [inline pkg/util/fedinformer/genericmanager/multi-cluster-manager.go:156]: The same problem.

File: pkg/search/backendstore/store.go [hunk 3]
```diff
@@ -61,13 +61,15 @@ func DeleteBackend(cluster string) {
...
-func GetBackend(cluster string) BackendStore {
+func GetBackend(cluster string, createDefault ...func(string) BackendStore) BackendStore {
 	backendLock.Lock()
 	defer backendLock.Unlock()
 	bs, ok := backends[cluster]
 	if !ok {
-		klog.Errorf("cannot find backend store %s, use defult backend", cluster)
-		return NewDefaultBackend(cluster)
+		if len(createDefault) > 0 && createDefault[0] != nil {
+			return createDefault[0](cluster)
```
[the remark this answers is not in the document]
Comment (tedli) [reply pkg/search/backendstore/store.go:70]: OK, I forget removing this here, it's needed to check return value is nil to add backend, not create a default one. I'm going to remove this.

File: pkg/search/controller.go [hunk 4]
```diff
@@ -161,46 +162,219 @@ func (c *Controller) handleErr(err error, key interface{}) {
...
+	}
+	for _, name := range added {
+		cr.registries[name] = struct{}{}
+		addedResources = c.getRegistryAddedResources(registries[name], cr, addedResources)
+	}
+	for _, name := range removed {
+		delete(cr.registries, name)
+		for resource, previousRegistries := range cr.resources {
+			delete(previousRegistries, name)
+			if len(previousRegistries) <= 0 {
```
[the remark this answers is not in the document]
Comment (tedli) [reply pkg/search/controller.go:217]: OK, it's just personal preference. `<=` here wants to tell something like `! >`. I'm going to correct this to meet community preference.

File: pkg/search/controller.go (hunk 1 above)
Comment (tedli) [reply pkg/search/controller.go:278]: OK, these is just for debugging during I code this pr, some of these log seems not necessary, I'm going to refactor logs related to this pr.

**Comment by XiShanYongYe-Chang** [comment]
Thanks @tedli @whitewindmills 
I'll take a review ASAP.

File: pkg/search/controller.go [hunk 5]
```diff
@@ -161,46 +162,219 @@ func (c *Controller) handleErr(err error, key interface{}) {
...
 
-// doCacheCluster processes the resourceRegistry object
-// TODO: update status
-func (c *Controller) doCacheCluster(cluster string) error {
-	// STEP0:  stop informer manager for the cluster which is not referenced by any `SearchRegistry` object.
-	v, ok := c.clusterRegistry.Load(cluster)
-	if !ok {
-		klog.Infof("Cluster %s is not registered", cluster)
-		return nil
+func (c *Controller) getClusterMatchedRegistries(cluster *clusterv1alpha1.Cluster) (indexedByName map[string]*searchv1alpha1.ResourceRegistry, matched []string, err error) {
```
[the remark this answers is not in the document]
Comment (tedli) [reply pkg/search/controller.go]: done

File: pkg/search/controller.go [hunk 6]
```diff
@@ -161,46 +162,219 @@ func (c *Controller) handleErr(err error, key interface{}) {
...
+func (c *Controller) reconcileClusterWithRegistries(cls *clusterv1alpha1.Cluster) (matchedRegistries map[string]*searchv1alpha1.ResourceRegistry, cr *clusterRegistry, resourcesChanged, newRegistry bool, err error) {
+	cluster := cls.Name
+	var allRegistries map[string]*searchv1alpha1.ResourceRegistry
+	var matched []string
+	allRegistries, matched, err = c.getClusterMatchedRegistries(cls)
+	if err != nil {
+		klog.Errorf("Get cluster matched resource registries failed, cluster: %s, error: %s", cluster, err)
+		return
+	}
+	if len(matched) > 0 {
```
[the remark this answers is not in the document]
Comment (tedli) [reply pkg/search/controller.go]: done

File: pkg/util/common.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
+package util
+
+import (
+	"fmt"
+	"strings"
+)
+
+func DiffKey[K comparable, V1, V2 any](previous map[K]V1, current map[K]V2) (added, removed []K) {
```
[the remark this answers is not in the document]
Comment (tedli) [reply pkg/util/common.go:11]: done

File: pkg/search/controller.go [hunk 8]
```diff
@@ -161,46 +163,210 @@ func (c *Controller) handleErr(err error, key interface{}) {
...
+		for resource, previousRegistries := range cr.resources {
+			delete(previousRegistries, name)
+			if len(previousRegistries) == 0 {
+				removedResources = append(removedResources, resource)
+			}
+		}
+		for _, resource := range removedResources {
+			delete(cr.resources, resource)
+		}
+	}
```
Comment (XiShanYongYe-Chang) [inline pkg/search/controller.go:226]: Do we need to handle deletions first before handling additions, because the added elements may contain deleted ones?

File: pkg/search/controller.go [hunk 9]
```diff
@@ -161,46 +163,210 @@ func (c *Controller) handleErr(err error, key interface{}) {
...
+	for _, registry := range matched {
+		addedResources = c.getRegistryAddedResources(registry, cr, addedResources)
+	}
```
Comment (XiShanYongYe-Chang) [inline pkg/search/controller.go:229]: This processing step feels redundant. All the matched RRs should have already been processed.

File: pkg/search/controller.go [hunk 10]
```diff
@@ -161,46 +163,210 @@ func (c *Controller) handleErr(err error, key interface{}) {
...
+	var addedRegistries, removedRegistries []string
+	if !hasExistingRegistries {
+		klog.Infof("Cluster %s has no existing registries", cluster)
+		addedRegistries, removedRegistries = util.DiffKey(registrySet(nil), matchedRegistries)
+	} else {
+		crv := v.(clusterRegistry)
+		cr = &crv
+		addedRegistries, removedRegistries = util.DiffKey(cr.registries, matchedRegistries)
+	}
+	if len(addedRegistries) > 0 {
```
Comment (XiShanYongYe-Chang) [inline pkg/search/controller.go:290]: Can we omit these judgments to improve the code readability and reduce cyclomatic complexity? Same as below.

File: pkg/search/controller.go [hunk 11]
```diff
@@ -161,46 +163,210 @@ func (c *Controller) handleErr(err error, key interface{}) {
...
+	if cr.unregistry() {
+		klog.Infof("Try to stop cluster informer %s", cluster)
+		c.InformerManager.Stop(cluster)
 		return nil
 	}
 
+	if resourcesChanged {
+		klog.Infof("Create new informer for cluster %s", cluster)
+		c.InformerManager.Stop(cluster)
+	}
```
Comment (XiShanYongYe-Chang) [inline pkg/search/controller.go:359]: Do we need to stop the informer before making changes to the resources, or can we directly modify the current informer?

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks~
Still reviewing.

File: pkg/search/controller.go [hunk 12]
```diff
@@ -274,58 +416,13 @@ func (c *Controller) updateResourceRegistry(oldObj, newObj interface{}) {
...
-		c.clusterRegistry.Store(cls, cr)
-
-		clusterSets[cls] = struct{}{}
-
-		// set backendstore
-		backendstore.AddBackend(cls, newRR.Spec.BackendStore)
-
-		c.queue.Add(cls)
+	clusterSet := sets.New[string]()
+	for _, registry := range []*searchv1alpha1.ResourceRegistry{newRR, oldRR} {
```
[the remark this answers is not in the document]
Comment (tedli) [reply pkg/search/controller.go:420]: This controller reconcile the relationship between clusters and registries, what's more the key of the controller's queue is cluster name.
At this point, we don't know cluster changed or not, the only thing we know is if `reflect.DeepEqual(oldRR.Spec, newRR.Spec)` not `true`, the registry must changed.
So it's need to do some reconcile. But, as I mentioned above, the key of the controller queue is cluster name, this leave us only 2 choices, 1 is to queue the cluster name, do reconcile at cluster level, that's what I did; 2 is to introduce another worker, which the key of queue is registry name, then we can reconcile at registry level.
However, if we decice to do using method 2, then we also have to retrieve clusters of old registry spec, and clusters of new registry spec, to find out what changed, then reconcile changed clusters, because finally reconciling multi cluster informer, is what we want.
It seemd not possible to reduce the redunancy, since if cluster-registry relationship not changed, the resources may also be changed, like previously we want to cache `Pod`, but currently we no longer want caching `Pod`, the cluster-registry relation not changed, but the resource did. That's why I use `diff` logic.

File: pkg/search/controller.go [hunk 13]
```diff
@@ -161,46 +163,210 @@ func (c *Controller) handleErr(err error, key interface{}) {
...
+			cr.resources[resource] = registrySet{name: struct{}{}}
+		}
+	}
+	return added
+}
+
+func (c *Controller) getClusterRegistriesModification(registries map[string]*searchv1alpha1.ResourceRegistry, matched map[string]*searchv1alpha1.ResourceRegistry, cr *clusterRegistry, added []string, removed []string) (updated *clusterRegistry, addedResources []schema.GroupVersionResource, removedResources []schema.GroupVersionResource) {
+	defer func() {
+		updated = cr
+	}()
```
[the remark this answers is not in the document]
Comment (tedli) [reply pkg/search/controller.go:205]: No, we not have to do this.

But by doing this we have 2 benefits, 1 is that the input cr arg could be `nil`; 2 is that at the callee level we know the cr may be changed by this function, which is more expressive.

That's some like `append`, `a = append(a, obj)`.

File: pkg/search/controller.go (hunk 8 above)
Comment (tedli) [reply pkg/search/controller.go:226]: I didn't got you, did you mean it's better to handle `removed` (L215) before `added` (L211)?

Either is ok.

File: pkg/search/controller.go (hunk 9 above)
Comment (tedli) [reply pkg/search/controller.go:229]: No, there may be added ones, and removed ones. But still remaining ones may have resources changed.
Like, registry target cluster is `aaa=bbb`, resources, `Pod`, `Node`, cluster label changed, and resources also changed, like adding `Event`.
So we not only need to handle registries added or removed, but also need to handle resources changes.

File: pkg/search/controller.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (XiShanYongYe-Chang) [reply pkg/search/controller.go:205]: ok, thx~

File: pkg/search/controller.go (hunk 10 above)
Comment (tedli) [reply pkg/search/controller.go:290]: Yes, at first I thought it's not need to stop the informer, just remove uneeded resources. So I did diff to find out what are added, what are removed.

But it's not work, we have to stop a informer if something changed, to init a new one. Unless we do a huge refactor of informer.

File: pkg/search/controller.go (hunk 11 above)
Comment (tedli) [reply pkg/search/controller.go:359]: No, like the comment above, I firstly came the same idea as you, so I did `diff` to find out what are added and what are removed. But it just not work.

You can find the first commit of this PR, I did the exact same thing, modifing the current informer.

File: pkg/search/controller.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (XiShanYongYe-Chang) [reply pkg/search/controller.go:420]: ok.

File: pkg/search/controller.go (hunk 11 above)
Comment (tedli) [reply pkg/search/controller.go:359]: https://github.com/karmada-io/karmada/blob/a9df8656a933d28e9f42e5c1a816ea9c9c1bde32/pkg/util/fedinformer/genericmanager/single-cluster-manager.go#L19-L22

Here, the first commit of this PR, I tried to add a `UnwatchResource(resource schema.GroupVersionResource)`, without success.

File: pkg/search/controller.go [hunk 14]
```diff
@@ -161,46 +163,210 @@ func (c *Controller) handleErr(err error, key interface{}) {
...
+	return
+}
 
-	cr := v.(clusterRegistry)
-	if cr.unregistry() {
-		klog.Infof("try to stop cluster informer %s", cluster)
-		c.InformerManager.Stop(cluster)
-		return nil
+func (c *Controller) getRegistryAddedResources(registry *searchv1alpha1.ResourceRegistry, cr *clusterRegistry, added []schema.GroupVersionResource) []schema.GroupVersionResource {
+	name := registry.Name
```
[the remark this answers is not in the document]
Comment (tedli) [reply pkg/search/controller.go]: done

File: pkg/search/controller.go (hunk 8 above)
Comment (XiShanYongYe-Chang) [reply pkg/search/controller.go:225]: Yes, what I mean is to delete first and then add, so as to prevent the elements from being cleared when deleting and adding elements that contain the same content.

File: pkg/search/controller.go (hunk 11 above)
Comment (XiShanYongYe-Chang) [reply pkg/search/controller.go:358]: ok, thx~

File: pkg/search/controller.go (hunk 8 above)
Comment (tedli) [reply pkg/search/controller.go]: done

File: pkg/util/common_test.go [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+		LatestVersion: "v0.9.2",
+	}
+	karmada = sigMultiCluster{
+		Name:          "karmada",
+		Status:        "active",
+		LatestVersion: "v1.6.0",
+	}
+)
+
+func TestDiffKey(t *testing.T) {
```
Comment (XiShanYongYe-Chang) [inline pkg/util/common_test.go:33]: Can this test achieve condition coverage? Consider the boundary conditions.

File: test/e2e/search_test.go [hunk 16]
```diff
@@ -710,5 +710,72 @@ var _ = ginkgo.Describe("[karmada-search] karmada search testing", ginkgo.Ordere
...
+				ginkgo.AfterAll(func() {
+					framework.RemoveResourceRegistry(karmadaClient, registryName)
+					framework.DeleteClusterLabels(karmadaClient, member1, clusterLabels)
+				})
+				ginkgo.It("cluster update trigger reconcile", func() {
+					// The cluster didn't match the label selector of registry
+					var proxyList *corev1.NodeList
+					gomega.Eventually(func(g gomega.Gomega) {
+						var err error
+						proxyList, err = proxyClient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
```
Comment (XiShanYongYe-Chang) [inline test/e2e/search_test.go:753]: The current modifications made in PR should not involve any changes in the search/proxy functionality. This test example may have been placed in the wrong location and can be moved to this side of line 344.

File: test/e2e/search_test.go [hunk 17]
```diff
@@ -710,5 +710,72 @@ var _ = ginkgo.Describe("[karmada-search] karmada search testing", ginkgo.Ordere
 				})
 			})
 		})
+		ginkgo.Describe("update cluster testing", ginkgo.Ordered, func() {
+			ginkgo.Context("cluster labeled after registry created", ginkgo.Ordered, func() {
```
Comment (XiShanYongYe-Chang) [inline test/e2e/search_test.go:714]: This line seems redundant.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thank you. Everything has been reviewed once, and only some testing issues remain.

File: pkg/util/common_test.go (hunk 15 above)
Comment (tedli) [reply pkg/util/common_test.go:40]: done

File: test/e2e/search_test.go (hunk 17 above)
Comment (tedli) [reply test/e2e/search_test.go:760]: I didn't touch this test case, it's provided by other commiter previously. It seemed that the diff algo of github get confused, it marks this line green treated as modification of this pr, but it's not.

File: pkg/util/common_test.go [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+		}
+		added, removed := DiffKey(previous, current)
+		if len(added) != 1 || added[0] != "karmada" {
+			t.Errorf("added = %v, want []string{\"karmada\"}", added)
+		}
+		if len(removed) != 1 || removed[0] != "kubefed" {
+			t.Errorf("removed = %v, want []string{\"kubefed\"}", removed)
+		}
+	})
+	return
```
[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/util/common_test.go:86]: echo lint errors here:
```
Installing golangci-lint v1.[5](https://github.com/karmada-io/karmada/actions/runs/5387138616/jobs/9778049938?pr=3670#step:5:6)2.2
golangci/golangci-lint info checking GitHub for tag 'v1.52.2'
golangci/golangci-lint info found version: 1.52.2 for v1.52.2/linux/amd[6](https://github.com/karmada-io/karmada/actions/runs/5387138616/jobs/9778049938?pr=3670#step:5:7)4
golangci/golangci-lint info installed /home/runner/go/bin/golangci-lint
Error: pkg/util/common_test.go:[8](https://github.com/karmada-io/karmada/actions/runs/5387138616/jobs/9778049938?pr=3670#step:5:9)6:2: S1023: redundant `return` statement (gosimple)
	return
	^

Please review the above warnings.
If the above warnings do not make sense, feel free to file an issue.
Error: Process completed with exit code 1.
```

File: test/e2e/search_test.go [hunk 19]
```diff
@@ -710,5 +756,42 @@ var _ = ginkgo.Describe("[karmada-search] karmada search testing", ginkgo.Ordere
 				})
 			})
 		})
+		ginkgo.Describe("update cluster testing", ginkgo.Ordered, func() {
```
Comment (XiShanYongYe-Chang) [inline test/e2e/search_test.go:759]: This new add test block can be removed.

File: test/e2e/search_test.go [hunk 20]
```diff
@@ -342,6 +342,52 @@ var _ = ginkgo.Describe("[karmada-search] karmada search testing", ginkgo.Ordere
 	})
 
 	ginkgo.Describe("backend store testing", ginkgo.Ordered, func() {
+
```
Comment (XiShanYongYe-Chang) [inline test/e2e/search_test.go:345]: How about setting this test case into a new `Describe`, just like:

```diff
--- a/test/e2e/search_test.go
+++ b/test/e2e/search_test.go
@@ -341,8 +341,7 @@ var _ = ginkgo.Describe("[karmada-search] karmada search testing", ginkgo.Ordere
                })
        })

-       ginkgo.Describe("backend store testing", ginkgo.Ordered, func() {
-
+       ginkgo.Describe("update cluster labels testing", ginkgo.Ordered, func() {
                var (
                        pollTimeout, pollInterval = time.Second * 10, time.Second
                        proxyClient               kubernetes.Interface
@@ -390,6 +389,9 @@ var _ = ginkgo.Describe("[karmada-search] karmada search testing", ginkgo.Ordere
                })
        })

+       ginkgo.Describe("backend store testing", ginkgo.Ordered, func() {
+       })
+
        ginkgo.Describe("karmada proxy testing", ginkgo.Ordered, func() {
                var (
                        pollTimeout, pollInterval       = time.Second * 10, time.Second
```

File: pkg/util/common_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (tedli) [reply pkg/util/common_test.go]: done

File: test/e2e/search_test.go (hunk 19 above)
Comment (tedli) [reply test/e2e/search_test.go]: done

File: test/e2e/search_test.go (hunk 20 above)
Comment (tedli) [reply test/e2e/search_test.go:348]: done

File: test/e2e/search_test.go [hunk 21]
```diff
@@ -344,6 +344,55 @@ var _ = ginkgo.Describe("[karmada-search] karmada search testing", ginkgo.Ordere
...
+			proxyClient               kubernetes.Interface
+			m1Dynamic                 dynamic.Interface
+			nodeGVR                   = corev1.SchemeGroupVersion.WithResource("nodes")
+		)
+
+		ginkgo.BeforeAll(func() {
+			m1Dynamic = framework.GetClusterDynamicClient(member1)
+			gomega.Expect(m1Dynamic).ShouldNot(gomega.BeNil())
+			proxyConfig := *restConfig
+			proxyConfig.Host += "/apis/search.karmada.io/v1alpha1/proxying/karmada/proxy"
```
Comment (XiShanYongYe-Chang) [inline test/e2e/search_test.go:360]: Hi @tedli, this e2e might not be written correctly. The test coverage does not include the modified content. In this revision, we have added handling for cluster modification events, but this only applies to search cache and not search proxy capability.

Have you tested the new processing logic locally? If it passes the local test, please share the test results. 
And then we can submit an E2E test for the modification in a subsequent PR to promote the integration of the current refactoring code.

Comment (RainbowMango) [reply test/e2e/search_test.go:360]: I'm saying that because this PR has been blocked for weeks.

Comment (XiShanYongYe-Chang) [reply test/e2e/search_test.go:360]: Yes, I think we can start by removing the E2E part and adding a separate PR to include e2e testing. This test can be completed by anyone who is interested. I create Issue #3731 to track it. How do you think @tedli

Comment (tedli) [reply test/e2e/search_test.go:360]: Hi all,
Sorry for late reply. I will remove e2e test case, and I'm glad to take a deeper dive with karmada e2e test, and then finish #3731

Comment (XiShanYongYe-Chang) [reply test/e2e/search_test.go:360]: Thanks @tedli

**Comment by XiShanYongYe-Chang** [comment]
I think we can merge now :)
Before merging it, can you help squash those commits into one commit?

**Comment by XiShanYongYe-Chang** [comment]
/lgtm
Ask an approve from @RainbowMango 
/cc @RainbowMango

**Review by RainbowMango — APPROVED** [review]
/approve
